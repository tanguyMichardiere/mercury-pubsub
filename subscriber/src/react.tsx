import type { ReactNode } from "react";
import { createContext, useContext, useEffect, useMemo } from "react";

import type { Channels } from "@mercury-pubsub/types";
import useEvent from "react-use-event-hook";

import type { SubscribeOptions } from ".";
import Subscriber from ".";

const subscriberContext = createContext<Subscriber | undefined>(undefined);

export default function SubscriberProvider(props: {
  url: string;
  _key: string;
  children: ReactNode;
}): JSX.Element {
  const subscriber = useMemo(() => new Subscriber(props.url, props._key), [props.url, props._key]);

  return (
    <subscriberContext.Provider value={subscriber}>{props.children}</subscriberContext.Provider>
  );
}

export function useSubscribe<C extends keyof Channels>(
  channel: C,
  events: Omit<SubscribeOptions<C>, "signal">
): void {
  const subscriber = useContext(subscriberContext);

  if (subscriber === undefined) {
    throw new Error("no subscriber context");
  }

  const abortController = useMemo(() => new AbortController(), []);

  // eslint-disable-next-line @typescript-eslint/no-empty-function
  const onopen = useEvent(events.onopen ?? async function () {});
  const ondata = useEvent(events.ondata);
  // eslint-disable-next-line @typescript-eslint/no-empty-function
  const onclose = useEvent(events.onclose ?? function () {});

  useEffect(
    function () {
      subscriber
        .subscribe(channel, { signal: abortController.signal, onopen, ondata, onclose })
        .catch(function (error) {
          console.error(error);
        });
      return function () {
        abortController.abort();
      };
    },
    [subscriber, channel, abortController, onopen, ondata, onclose]
  );
}
