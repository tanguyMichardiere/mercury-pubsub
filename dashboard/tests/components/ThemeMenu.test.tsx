import { fireEvent, render, screen } from "@testing-library/react";
import renderer from "react-test-renderer";

import ThemeMenu from "../../src/components/ThemeMenu";

const setTheme = jest.fn();
jest.mock("next-themes", () => ({
  useTheme: jest.fn(() => ({ setTheme })),
}));

describe("ThemeMenu", function () {
  it("renders unchanged", function () {
    const tree = renderer.create(<ThemeMenu />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("opens correctly", function () {
    render(<ThemeMenu />);
    fireEvent.click(screen.getByText(/theme/i));
    expect(screen.getByText(/light/i)).toBeDefined();
    expect(screen.getByText(/dark/i)).toBeDefined();
    expect(screen.getByText(/system/i)).toBeDefined();
  });

  it("changes themes correctly", function () {
    render(<ThemeMenu />);
    const menuButton = screen.getByText(/theme/i);

    fireEvent.click(menuButton);
    fireEvent.click(screen.getByText(/light/i));
    expect(setTheme).toHaveBeenCalledTimes(1);
    expect(setTheme).toHaveBeenLastCalledWith("light");

    fireEvent.click(menuButton);
    fireEvent.click(screen.getByText(/dark/i));
    expect(setTheme).toHaveBeenCalledTimes(2);
    expect(setTheme).toHaveBeenLastCalledWith("dark");

    fireEvent.click(menuButton);
    fireEvent.click(screen.getByText(/system/i));
    expect(setTheme).toHaveBeenCalledTimes(3);
    expect(setTheme).toHaveBeenLastCalledWith("system");
  });

  it("styles menu items correctly on hover", function () {
    render(<ThemeMenu />);
    fireEvent.click(screen.getByText(/theme/i));
    const lightMenuItem = screen.getByText(/light/i);
    const darkMenuItem = screen.getByText(/dark/i);
    const systemMenuItem = screen.getByText(/system/i);

    fireEvent.focus(lightMenuItem);
    expect(lightMenuItem.classList).toContain("bg-violet-500");
    expect(darkMenuItem.classList).not.toContain("bg-violet-500");
    expect(systemMenuItem.classList).not.toContain("bg-violet-500");

    fireEvent.focus(darkMenuItem);
    expect(lightMenuItem.classList).not.toContain("bg-violet-500");
    expect(darkMenuItem.classList).toContain("bg-violet-500");
    expect(systemMenuItem.classList).not.toContain("bg-violet-500");

    fireEvent.focus(systemMenuItem);
    expect(lightMenuItem.classList).not.toContain("bg-violet-500");
    expect(darkMenuItem.classList).not.toContain("bg-violet-500");
    expect(systemMenuItem.classList).toContain("bg-violet-500");
  });
});
