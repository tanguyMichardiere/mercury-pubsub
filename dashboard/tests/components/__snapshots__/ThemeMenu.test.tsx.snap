// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ThemeMenu renders unchanged 1`] = `
<div
  className="text-right"
>
  <div
    className="relative inline-block text-left"
  >
    <div>
      <button
        aria-expanded={false}
        aria-haspopup={true}
        className="inline-flex w-full justify-center rounded-md bg-white px-4 py-2 text-sm font-medium shadow-md dark:bg-gray-700"
        id="headlessui-menu-button-:r0:"
        onClick={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        type="button"
      >
        Theme
        <svg
          aria-hidden="true"
          className="ml-2 -mr-1 h-5 w-5"
          fill="currentColor"
          viewBox="0 0 20 20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clipRule="evenodd"
            d="M5.293 7.293a1 1 0 011.414 0L10 10.586l3.293-3.293a1 1 0 111.414 1.414l-4 4a1 1 0 01-1.414 0l-4-4a1 1 0 010-1.414z"
            fillRule="evenodd"
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;
